### ISSUE TITLE
`#[expect(dead_code)]` does not behave identically to `#[allow(dead_code)]`

### ISSUE THREAD

**Issue by ojeda** [rust-lang/rust#114557]
In its [RFC](https://rust-lang.github.io/rfcs/2383-lint-reasons.html#expect-lint-attribute), the [`#[expect(...)]` lint attribute](https://github.com/rust-lang/rust/issues/54503) is said to function identically to `#[allow(...)]`. However, when trying out the attribute in the kernel, I found a case where I can see a difference.

The following code does not emit any warning:

```rust
fn f() {}

#[allow(dead_code)]
fn g() {
    f();
}
```

However, this does:

```rust
fn f() {}

#[expect(dead_code)]
fn g() {
    f();
}
```

In other words, with `#[allow(...)]`, `f` is understood to be used, because it is called from `g`, even if `g` is dead code itself. With `#[expect(...)]`, `f` is understood as not used.

Both 1.71.1 and nightly (32303b219 2023-07-29) behave the same way: https://godbolt.org/z/sMhYh1q11.

Cc'ing @xFrednet and @fmease who may be interested in this.

**Comment by xFrednet** [comment]
Interesting, I assume that this is due to the actual lint implementation. It probably somewhere checks for the lint level and has a `== Allow` in it.

**Comment by ojeda** [comment]
Does it only happen with nightly 1.85 and Edition 2024, or did you also try other versions/editions? Could you please reduce it to a small example if possible?

Thanks!
